compositor_render/src/renderer/scene/constraints.rs

```diff
@@ -0,0 +1,46 @@
+use compositor_common::{
+    error::ConstraintsValidationError,
+    scene::{
+        validation::{constraints::NodeConstraints, inputs::InputsCountConstraint},
+        NodeParams, NodeSpec, SceneSpec,
+    },
+};
+
+pub fn validate_constraints(scene: &SceneSpec) -> Result<(), ConstraintsValidationError> {
+    for (node_id, node_constraints) in scene
+        .nodes
+        .iter()
+        .map(|node| (&node.node_id, node_constraints(node)))
+    {
+        node_constraints.validate(scene, node_id)?
+    }
+
+    Ok(())
+}
+
+fn node_constraints(node: &NodeSpec) -> NodeConstraints {
+    // TODO: make web renderer and shader constraints API configurable
```

Comment:
I would prefer not to have any node-specific code here. I would either:
- move the entire function to  NodeSpec or NodeParams
- leave a match statement here, but only call to some method on each of those types.

This file should not contain any logic specific to a type

Reply:
In the case of user-defined constraints, we need to access them somehow in runtime. We can't move into `comspoitor_common`, since this function will require access to some renderer contexts, and `compositor_common` shouldn't depend on `compositor_renderer`